File: solutions/cpp/interest-is-interesting/1/interest_is_interesting.cpp

struct DollarAndInterest {
    double amount;
    double interest;
};





// interest_rate returns the interest rate for the provided balance.
double interest_rate(double balance) {
    double interest{0.0};

    if (balance < 0) {
        interest = 3.213;
    } else if (balance < 1000.0) {
        interest = 0.5;
    } else if (balance < 5000.0) {
        interest = 1.621;
    } else {
        interest = 2.475;
    }

    return interest;
}

// yearly_interest calculates the yearly interest for the provided balance.
double yearly_interest(double balance) {
    double accruedInterest{0.0};
    accruedInterest = interest_rate(balance) * 0.01 * balance;
    
    return accruedInterest;
}

// annual_balance_update calculates the annual balance update, taking into
// account the interest rate.
double annual_balance_update(double balance) {
    double yearEndBalance{0.0};
    yearEndBalance = yearly_interest(balance) + balance;
    return yearEndBalance;
}

// years_until_desired_balance calculates the minimum number of years required
// to reach the desired balance.
int years_until_desired_balance(double balance, double target_balance) {
    double currentBalance{balance};
    int yearsToAchieve{0};

    for (; currentBalance < target_balance; ++yearsToAchieve) {
        currentBalance = annual_balance_update(currentBalance);
    }

    return yearsToAchieve;




    
    return 0;
}

File: solutions/go/interest-is-interesting/1/interest_is_interesting.go
package interest

// InterestRate returns the interest rate for the provided balance.
func InterestRate(balance float64) float32 {
	ratePercent := 0.0
	if balance < 0 {
		ratePercent = 3.213
	} else if balance < 1000.0 {
		ratePercent = 0.5
	} else if balance < 5000.0 {
		ratePercent = 1.621
	} else {
		ratePercent = 2.475
	}

	return float32(ratePercent)
}

// Interest calculates the interest for the provided balance.
func Interest(balance float64) float64 {
	rate := float64(InterestRate(balance))
	return balance * (rate / 100.0)
}

// AnnualBalanceUpdate calculates the annual balance update, taking into account the interest rate.
func AnnualBalanceUpdate(balance float64) float64 {
	return balance + Interest(balance)
}

// YearsBeforeDesiredBalance calculates the minimum number of years required to reach the desired balance.
func YearsBeforeDesiredBalance(balance, targetBalance float64) int {
	years := 0
	currentBalance := balance
	for currentBalance < targetBalance {
		years++
		currentBalance = AnnualBalanceUpdate(currentBalance)
	}
	return years
}
